src/sync/sync_manager.cpp

```diff
@@ -313,28 +311,82 @@ std::shared_ptr<SyncUser> SyncManager::get_user(const std::string& identity,
     }
 }
 
-std::shared_ptr<SyncUser> SyncManager::get_existing_logged_in_user(const std::string& identity) const
+std::shared_ptr<SyncUser> SyncManager::get_admin_token_user_from_identity(const std::string& identity,
+                                                                          util::Optional<std::string> server_url,
+                                                                          util::Optional<std::string> token)
 {
+    if (server_url)
+        return get_admin_token_user(*server_url, std::move(token), identity);
+
     std::lock_guard<std::mutex> lock(m_user_mutex);
-    auto it = m_users.find(identity);
-    if (it == m_users.end()) {
-        return nullptr;
+    // Look up the user based off the identity.
+    // No server URL, so no migration possible.
+    auto it = m_admin_token_users.find(identity);
+    if (it == m_admin_token_users.end()) {
+        // No existing user.
+        if (!token)
+            throw std::invalid_argument("User did not exist, but token was not provided.");
+
+        auto new_user = std::make_shared<SyncUser>(std::move(*token),
+                                                   c_admin_identity,
+                                                   std::move(server_url),
+                                                   identity,
+                                                   SyncUser::TokenType::Admin);
+        m_admin_token_users.insert({ identity, new_user });
+        return new_user;
+    } else {
+        return it->second;
     }
-    auto ptr = it->second;
-    return (ptr->state() == SyncUser::State::Active ? ptr : nullptr);
+}
+
+std::shared_ptr<SyncUser> SyncManager::get_admin_token_user(const std::string& server_url,
+                                                            util::Optional<std::string> token,
+                                                            util::Optional<std::string> old_identity)
+{
+    std::shared_ptr<SyncUser> user;
+    {
+        std::lock_guard<std::mutex> lock(m_user_mutex);
+        // Look up the user based off the server URL.
+        auto it = m_admin_token_users.find(server_url);
+        if (it == m_admin_token_users.end()) {
+            // No existing user.
+            if (!token)
+                throw std::invalid_argument("User did not exist, but token was not provided.");
+
+            auto new_user = std::make_shared<SyncUser>(std::move(*token),
+                                                       c_admin_identity,
+                                                       server_url,
+                                                       c_admin_identity + server_url,
+                                                       SyncUser::TokenType::Admin);
+            m_admin_token_users.insert({ server_url, new_user });
+            user = std::move(new_user);
+        } else {
+            user = it->second;
```

Comment:
If the user already exists, we probably don't want to be renaming the user directory out from under it since there might already be active sync sessions.

My suggestion would be to invert the `if` condition and turn the "user found" case into an early return in the true branch, then handle the "user not found" case after the `if` statement.